presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java

```diff
@@ -90,14 +100,55 @@ private static boolean shouldRewriteTableLayout(TableScanNode source)
         @Override
         public Result apply(FilterNode filterNode, Captures captures, Context context)
         {
+            TableScanNode tableScan = captures.get(TABLE_SCAN);
+
             TableLayoutRewriter tableLayoutRewriter = new TableLayoutRewriter(metadata, context.getSession(), context.getSymbolAllocator(), context.getIdAllocator());
             PlanNode rewrittenTableScan = tableLayoutRewriter.planTableScan(captures.get(TABLE_SCAN), filterNode.getPredicate());
 
-            if (rewrittenTableScan instanceof TableScanNode || rewrittenTableScan instanceof ValuesNode || (((FilterNode) rewrittenTableScan).getPredicate() != filterNode.getPredicate())) {
-                return Result.ofPlanNode(rewrittenTableScan);
+            if (arePlansEquivalent(filterNode, tableScan, rewrittenTableScan, context)) {
+                return Result.empty();
+            }
+
+            return Result.ofPlanNode(rewrittenTableScan);
+        }
+
+        public boolean arePlansEquivalent(FilterNode filter, TableScanNode tableScan, PlanNode rewritten, Context context)
+        {
+            Expression resultPredicate = BooleanLiteral.TRUE_LITERAL;
+            if (rewritten instanceof FilterNode) {
+                resultPredicate = ((FilterNode) rewritten).getPredicate();
+            }
+
+            Optional<TableScanNode> rewrittenTableScan = searchFrom(rewritten)
+                    .where(TableScanNode.class::isInstance)
+                    .findSingle();
```

Comment:
If `rewrittenTableScan` is anything different than just `(TableScanNode ) rewritten.getSource()` then:
- the rewritten plan is definitely different (the rule matches filter directly above table scan)
- you won't notice it

Replace `searchFrom()` with direct `instanceof` check of `rewritten.getSource()`.